File: src/android/org/apache/cordova/plugin/Wl_Pay_Ccr_DirectConnect.java
package org.apache.cordova.plugin;

import android.Manifest;
import android.content.Context;

import com.directconnect.mobilesdk.device.AugustaDeviceManager;
import com.directconnect.mobilesdk.device.BTMagDeviceManager;
import com.directconnect.mobilesdk.device.CardData;
import com.directconnect.mobilesdk.device.Device;
import com.directconnect.mobilesdk.device.DeviceManager;
import com.directconnect.mobilesdk.device.MagtekDeviceManager;
import com.directconnect.mobilesdk.device.MiuraDeviceManager;
import com.directconnect.mobilesdk.device.PINData;
import com.directconnect.mobilesdk.device.UniMagDeviceManager;
import com.directconnect.mobilesdk.device.VirtualDeviceManager;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Connector for mobile SDK of NMI.
 */
public class Wl_Pay_Ccr_DirectConnect extends Wl_Pay_Ccr_Abstract implements DeviceManager.Listener
{
  /**
   * A list of devices.
   *
   * Result of <tt>XXXDeviceManager.getAvailableDevices();</tt>.
   */
  private Device[] devices;

  /**
   * Device manager object.
   */
  private DeviceManager deviceManager;

  /**
   * ID of device that is currently loaded.
   */
  private int id_device;

  private CardData o_card_last=null;

  /**
   * Creates a new Direct Connect interface object.
   *
   * @return Direct Connect interface object.
   */
  public static Wl_Pay_Ccr_DirectConnect create(Wl_Pay_Ccr o_controller) throws JSONException, Wl_UserException
  {
    o_controller.logInfo("Config: "+o_controller.config().toString());
    JSONObject a_config=o_controller.config().getJSONObject("a_processor");

    int id_device=a_config.getInt("id_device");
    Device[] devices;
    DeviceManager deviceManager;
    Context o_context=o_controller.cordova.getActivity().getApplicationContext();

    switch (id_device)
    {
      case Wl_DeviceSid.DC_IDT_AUGUSTA:
        devices = AugustaDeviceManager.getAvailableDevices();
        if(devices==null)
          throw new Wl_UserException("augusta-null","Can not initialize Augusta Device Manager (AugustaDeviceManager.getAvailableDevices() returns null).");
        if(devices.length==0)
          throw new Wl_UserException("augusta-empty","Can not initialize Augusta Device Manager (AugustaDeviceManager.getAvailableDevices() returns an empty array).");
        deviceManager = new AugustaDeviceManager(devices[0], o_context);
        break;
      case Wl_DeviceSid.DC_IDT_BT_MAG:
        devices = BTMagDeviceManager.getAvailableDevices();
        //noinspection ConstantConditions
        if(devices==null)
          throw new Wl_UserException("bt-mag-null","Can not initialize IDTech BtMag Device Manager (BTMagDeviceManager.getAvailableDevices() returns null).");
        if(devices.length==0)
          throw new Wl_UserException("bt-mag-empty","Can not initialize IDTech BtMag Device Manager (BTMagDeviceManager.getAvailableDevices() returns an empty array).");
        deviceManager = new BTMagDeviceManager(devices[0], o_context);
        break;
      case Wl_DeviceSid.DC_IDT_UNI_MAG:
        devices = UniMagDeviceManager.getAvailableDevices();
        if(devices.length==0)
          throw new Wl_UserException("uni-mag-empty","Can not initialize IDTech UniMag Device Manager (UniMagDeviceManager.getAvailableDevices() returns an empty array).");
        deviceManager = new UniMagDeviceManager(devices[0], o_context);
        break;
// This became unavailable after update on 2018-05-07 where we added Magtek.
// If you uncomment this, uncomment also counterpart code in debugInfo() in this file.
//      case Wl_DeviceSid.DC_IDT_UNI_PAY:
//        devices = UniPayDeviceManager.getAvailableDevices();
//        if(devices.length==0)
//          throw new Wl_UserException("uni-pay-empty","Can not initialize IDTech UniPay Device Manager (UniPayDeviceManager.getAvailableDevices() returns an empty array).");
//        deviceManager = new UniPayDeviceManager(devices[0], o_context);
//        break;
      case Wl_DeviceSid.DC_MAGTEK_AUDIO:
        devices = MagtekDeviceManager.getAvailableDevices();
        //noinspection ConstantConditions
        if(devices==null)
          throw new Wl_UserException("magtek-null","Can not initialize Magtek Device Manager (MagtekDeviceManager.getAvailableDevices() returns null).");
        if(devices.length==0)
          throw new Wl_UserException("magtek-empty","Can not initialize Magtek Device Manager (MagtekDeviceManager.getAvailableDevices() returns an empty array).");
        switch (id_device)
        {
          case Wl_DeviceSid.DC_MAGTEK_AUDIO:
            deviceManager = new MagtekDeviceManager(devices[0], o_context,"Audio");
            break;
          default:
            throw new Wl_UserException("magtek-internal","Internal plugin error (Magtek connection type is not implemented).");
        }
        break;
      case Wl_DeviceSid.DC_MIURA:
        devices = MiuraDeviceManager.getAvailableDevices();
        //noinspection ConstantConditions
        if(devices==null)
          throw new Wl_UserException("miura-null","Can not initialize Miura Device Manager (MiuraDeviceManager.getAvailableDevices() returns null).");
        if(devices.length==0)
          throw new Wl_UserException("miura-empty","Can not initialize Miura Device Manager (MiuraDeviceManager.getAvailableDevices() returns an empty array).");
        deviceManager = new MiuraDeviceManager(devices[0], o_context);
        break;
      case Wl_DeviceSid.VIRTUAL:
        devices = VirtualDeviceManager.getAvailableDevices();
        if(devices.length==0)
          throw new Wl_UserException("virtual-empty","Can not initialize Virtual Device Manager (VirtualDeviceManager.getAvailableDevices() returns an empty array).");
        deviceManager = new VirtualDeviceManager(devices[0], o_controller.cordova.getActivity());
        break;
      default:
        throw new Wl_UserException("dc-not-implemented","Interface for this DirectConnect device is not implemented.");
    }

    Wl_Pay_Ccr_DirectConnect o_result=new Wl_Pay_Ccr_DirectConnect();
    o_result.devices=devices;
    o_result.deviceManager=deviceManager;
    o_result.id_device=id_device;
    return o_result;
  }

  @Override
  JSONObject debugGet() throws JSONException
  {
    JSONObject a_debug=new JSONObject();

    JSONArray a_device=new JSONArray();
    for (Device device : this.devices)
    {
      JSONObject a_device_item=new JSONObject();
      a_device_item.put("address",device.getAddress());
      a_device_item.put("name",device.getName());
      a_device_item.put("type",device.getType());
      a_device_item.put("class",device.getClass().getName());
      a_device.put(a_device_item);
    }

    a_debug.put("devices",a_device);
    a_debug.put("devices.length",this.devices==null?"[null]":this.devices.length);
    a_debug.put("deviceManager.class",this.deviceManager==null?"[null]":this.deviceManager.getClass().getSimpleName());
    a_debug.put("id_device",this.id_device);
    a_debug.put("this.class",this.getClass().getSimpleName());

    if(this.deviceManager!=null)
    {
      a_debug.put("deviceManager.isCardInserted",this.deviceManager.isCardInserted());
      a_debug.put("deviceManager.isConnected",this.deviceManager.isConnected());
    }

    if(this.o_card_last==null)
      a_debug.put("card","[null]");
    else
    {
      JSONObject a_card=new JSONObject();
      a_card.put("getCardholderName",this.o_card_last.getCardholderName());
      a_card.put("getDataBlock",this.o_card_last.getDataBlock());
      a_card.put("getDataType",this.o_card_last.getDataType().toString());
      a_card.put("getExpDate",this.o_card_last.getExpDate());
      a_card.put("getKSN",this.o_card_last.getKSN());
      a_card.put("getPAN",this.o_card_last.getPAN());
      a_card.put("getServiceCode",this.o_card_last.getServiceCode());
      a_card.put("getTrack1",this.o_card_last.getTrack1());
      a_card.put("getTrack2",this.o_card_last.getTrack2());
      a_card.put("getTrack3",this.o_card_last.getTrack3());

      a_debug.put("card",a_card);
    }

    return a_debug;
  }

  static JSONObject debugGlobal() throws JSONException
  {
    JSONObject a_debug=new JSONObject();

    Device[] devices;

    devices = AugustaDeviceManager.getAvailableDevices();
    if(devices==null)
      a_debug.put("AugustaDeviceManager.getAvailableDevices","[null]");
    else if(devices.length==0)
      a_debug.put("AugustaDeviceManager.getAvailableDevices","[empty array]");
    else
      a_debug.put("AugustaDeviceManager.getAvailableDevices",devices.length);

    devices = BTMagDeviceManager.getAvailableDevices();
    //noinspection ConstantConditions
    if(devices==null)
      a_debug.put("BTMagDeviceManager.getAvailableDevices","[null]");
    else if(devices.length==0)
      a_debug.put("BTMagDeviceManager.getAvailableDevices","[empty array]");
    else
      a_debug.put("BTMagDeviceManager.getAvailableDevices",devices.length);

    devices = UniMagDeviceManager.getAvailableDevices();
    if(devices.length==0)
      a_debug.put("UniMagDeviceManager.getAvailableDevices","[empty array]");
    else
      a_debug.put("UniMagDeviceManager.getAvailableDevices",devices.length);

//    devices = UniPayDeviceManager.getAvailableDevices();
//    if(devices.length==0)
//      a_debug.put("UniPayDeviceManager.getAvailableDevices","[empty array]");
//    else
//      a_debug.put("UniPayDeviceManager.getAvailableDevices",devices.length);

    devices = MagtekDeviceManager.getAvailableDevices();
    //noinspection ConstantConditions
    if(devices==null)
      a_debug.put("MagtekDeviceManager.getAvailableDevices","[null]");
    else if(devices.length==0)
      a_debug.put("MagtekDeviceManager.getAvailableDevices","[empty array]");
    else
      a_debug.put("MagtekDeviceManager.getAvailableDevices",devices.length);

    devices = MiuraDeviceManager.getAvailableDevices();
    //noinspection ConstantConditions
    if(devices==null)
      a_debug.put("MiuraDeviceManager.getAvailableDevices","[null]");
    else if(devices.length==0)
      a_debug.put("MiuraDeviceManager.getAvailableDevices","[empty array]");
    else
      a_debug.put("MiuraDeviceManager.getAvailableDevices",devices.length);

    devices = VirtualDeviceManager.getAvailableDevices();
    if(devices.length==0)
      a_debug.put("VirtualDeviceManager.getAvailableDevices","[empty array]");
    else
      a_debug.put("VirtualDeviceManager.getAvailableDevices",devices.length);

    return a_debug;
  }

  @Override
  public void onCardInserted(CardData cardData)
  {
    try
    {
      this.logInfo("[Wl_Pay_Ccr_DirectConnect.onCardInserted]");
    }
    catch (JSONException ignored)
    {
    }
  }

  @Override
  public void onCardRemoved()
  {
    try
    {
      this.logInfo("[Wl_Pay_Ccr_DirectConnect.onCardRemoved]");
    }
    catch (JSONException ignored)
    {
    }
  }

  /**
   * Receives card swiped event.
   *
   * @param cardData Populated <tt>CardData</tt> object. A <tt>null</tt> value indicates the
   *   operation was timed out or was cancelled by user. A non-null <tt>CardData</tt> with
   *   <tt>DataType.nil</tt> indicates a bad swipe.
   */
  @Override
  public void onCardSwiped(CardData cardData)
  {
    try
    {
      this.o_card_last=cardData;

      if(cardData!=null&&cardData.getDataType()!=CardData.DataType.nil&&cardData.getEncryptionParameters()!=null)
      {
        JSONObject a_card=new JSONObject();

        JSONObject a_encrypt=new JSONObject();
        a_encrypt.put("DataBlock",cardData.getDataBlock()); // Example: "F1DAC156A909552E...E8E62EA1AA103906"
        a_encrypt.put("EncryptionType", cardData.getEncryptionParameters().getEncryptionType()); // Example: "DUKPT"
        a_encrypt.put("HSMDevice", cardData.getEncryptionParameters().getHSMDevice()); // Example: "Thales"
        a_encrypt.put("KSN",cardData.getKSN()); // Example: "0000020000004FA00001"
        a_encrypt.put("TerminalType", cardData.getEncryptionParameters().getTerminalType()); // Example: "Miura"

        a_card.put("a_encrypt",a_encrypt);
        a_card.put("s_number_mask",cardData.getPAN()); // Example: "450220******1234"
        a_card.put("s_expire",cardData.getExpDate()); // Example: "0318"
        a_card.put("s_device","android.dc."+this.deviceManager.getClass().getName());
        // a_card.put("s_holder",cardData.getCardholderName()); Returns null.

        // a_card.put("s_track_1",cardData.getTrack1()); // Returns null.
        // a_card.put("s_track_2",cardData.getTrack2()); // contains masked credit card a expiration date, and not a valid track value.
        // a_card.put("s_track_3",cardData.getTrack3()); // Returns null.

        // a_card.put("getDataType",this.o_card_last.getDataType().toString()); // Returns: "P2PE"
        // a_card.put("getServiceCode",cardData.getServiceCode()); // Example: 201

        this.controller().fireSwipe(a_card);
      }
      else
        this.controller().fireSwipeError();

      if(this.id_device!=Wl_DeviceSid.VIRTUAL)
        this.deviceManager.acceptCard("Swipe Card");
    }
    catch (Exception e)
    {
      this.controller()._exception(null,e);
    }
  }

  @Override
  public void onConnected()
  {
    try
    {
      this.logInfo("[Wl_Pay_Ccr_DirectConnect.onConnected]");
      if(this.id_device!=Wl_DeviceSid.VIRTUAL)
        this.deviceManager.acceptCard("Swipe Card");
    }
    catch (JSONException ignored)
    {
    }
  }

  @Override
  public void onDisconnected()
  {
    try
    {
      this.logInfo("[Wl_Pay_Ccr_DirectConnect.onDisconnected]");
    }
    catch (JSONException ignored)
    {
    }
  }

  @Override
  public void onMenuSelected(int i)
  {
    try
    {
      this.logInfo("[Wl_Pay_Ccr_DirectConnect.onMenuSelected]");
    }
    catch (JSONException ignored)
    {
    }
  }

  @Override
  public void onPINEntered(PINData pinData)
  {
    try
    {
      this.logInfo("[Wl_Pay_Ccr_DirectConnect.onPINEntered]");
    }
    catch (JSONException ignored)
    {
    }
  }

  @Override
  public void onYNAnswered(int i)
  {
    try
    {
      this.logInfo("[Wl_Pay_Ccr_DirectConnect.onYNAnswered]");
    }
    catch (JSONException ignored)
    {
    }
  }

  @Override
  protected String[] permissionList() throws JSONException
  {
    switch (this.id_device)
    {
      case Wl_DeviceSid.DC_IDT_AUGUSTA:
      case Wl_DeviceSid.DC_IDT_BT_MAG:
      case Wl_DeviceSid.DC_IDT_UNI_MAG:
      case Wl_DeviceSid.DC_IDT_UNI_PAY:
      case Wl_DeviceSid.DC_MAGTEK_AUDIO:
        // Union of all privileges from these files:
        // \src\android\dc\DCMobileSDK.aar\AndroidManifest.xml
        // \src\android\dc\DCMobileSDK-IDT.aar\AndroidManifest.xml
        // Equals content of
        // \src\android\dc\DCMobileSDK-Magtek.aar\AndroidManifest.xml
        return new String[]{
          Manifest.permission.MODIFY_AUDIO_SETTINGS,
          Manifest.permission.RECORD_AUDIO,
          // Manifest.permission.MOUNT_UNMOUNT_FILESYSTEMS,
          Manifest.permission.WRITE_EXTERNAL_STORAGE,
          Manifest.permission.READ_EXTERNAL_STORAGE,
          Manifest.permission.INTERNET,
          Manifest.permission.ACCESS_NETWORK_STATE,
          Manifest.permission.BLUETOOTH,
          Manifest.permission.BLUETOOTH_ADMIN
        };
      case Wl_DeviceSid.DC_MIURA:
        // Union of all privileges from these files:
        // \src\android\dc\DCMobileSDK.aar\AndroidManifest.xml
        // \src\android\dc\DCMobileSDK-Miura.aar\AndroidManifest.xml
        // \src\android\dc\Lib-Miura-SDK.aar\AndroidManifest.xml
        return new String[]{
          Manifest.permission.ACCESS_WIFI_STATE,
          Manifest.permission.BLUETOOTH,
          Manifest.permission.BLUETOOTH_ADMIN,
          Manifest.permission.GET_TASKS,
          Manifest.permission.INTERNET,
          Manifest.permission.WRITE_EXTERNAL_STORAGE
        };
      case Wl_DeviceSid.VIRTUAL:
        return new String[]{};
      default:
        this.logError("[Wl_Pay_Ccr_DirectConnect.permissionList] Device ID is not known.");
        return new String[]{};
    }
  }

  @Override
  public void startup() throws JSONException
  {
    this.logInfo("[Wl_Pay_Ccr_DirectConnect.startup]");
    this.deviceManager.connect(this);
  }

  @Override
  public void tearDown()
  {
    if(this.deviceManager!=null)
    {
      this.deviceManager.disconnect();
      this.deviceManager=null;
    }
    this.devices=null;
  }

  @Override
  public void testSwipe(JSONObject a_card) throws JSONException
  {
    if(this.deviceManager==null)
    {
      this.logError("deviceManager is not initialized.");
      return;
    }

    if(this.id_device!=Wl_DeviceSid.VIRTUAL)
    {
      this.logError("Can not do testSwipe() because current device is not for testing purposes.");
      return;
    }

    this.deviceManager.acceptCard("Enter card data.");
  }
}

File: src/ios/Wl_Pay_Ccr_DirectConnect.m
#import "Wl_Pay_Ccr_DirectConnect.h"
#import "Wl_DeviceSid.h"
#import "Wl_UserException.h"

@implementation Wl_Pay_Ccr_DirectConnect

+ (Wl_Pay_Ccr_DirectConnect*)create: (Wl_Pay_Ccr*)o_controller
{
    [o_controller logInfo:[NSString stringWithFormat:@"Config: %@",[o_controller config]]];

    NSDictionary* a_config = [[o_controller config] objectForKey:@"a_processor"];

    short id_device = [[a_config objectForKey:@"id_device"] shortValue];
    NSArray* devices;
    DCGDeviceManager* deviceManager;

    switch(id_device)
    {
        /* Not implemented for iOS.
         case WL_DEVICE_DC_IDT_AUGUSTA:
            devices = AugustaDeviceManager.getAvailableDevices();
            if(devices==null||devices.length==0)
                return null;
            deviceManager = new AugustaDeviceManager(devices[0], o_context);
            break;*/
        /*case WL_DEVICE_DC_IDT_BT_MAG:
            devices = [DCGBTMagDeviceManager getAvailableDevices];
            if(devices==nil)
                @throw [Wl_UserException exceptionWithName:@"bt-mag-null" reason:@"Can not initialize IDTech BtMag Device Manager ([DCGBTMagDeviceManager getAvailableDevices] returns null)." userInfo:nil];
            if([devices count]==0)
                @throw [Wl_UserException exceptionWithName:@"bt-mag-empty" reason:@"Can not initialize IDTech BtMag Device Manager ([DCGBTMagDeviceManager getAvailableDevices] returns an empty array)." userInfo:nil];
            deviceManager = [[DCGBTMagDeviceManager alloc] init:[devices objectAtIndex:0]];
            break;
        case WL_DEVICE_DC_IDT_UNI_MAG:
            devices = [DCGUniMagDeviceManager getAvailableDevices];
            if(devices==nil)
                @throw [Wl_UserException exceptionWithName:@"uni-mag-null" reason:@"Can not initialize IDTech UniMag Device Manager ([DCGUniMagDeviceManager getAvailableDevices] returns null)." userInfo:nil];
            if([devices count]==0)
                @throw [Wl_UserException exceptionWithName:@"uni-mag-empty" reason:@"Can not initialize IDTech UniMag Device Manager ([DCGUniMagDeviceManager getAvailableDevices] returns an empty array)." userInfo:nil];
            deviceManager = [[DCGUniMagDeviceManager alloc] init:[devices objectAtIndex:0]];
            break;*/
        case WL_DEVICE_DC_MAGTEK_AUDIO:
            devices = [DCGUADynamoDeviceManager getAvailableDevices];
            if(devices==nil)
                @throw [Wl_UserException exceptionWithName:@"magtek-null" reason:@"Can not initialize Magtek Device Manager ([DCGUADynamoDeviceManager getAvailableDevices] returns null)." userInfo:nil];
            if([devices count]==0)
                @throw [Wl_UserException exceptionWithName:@"magtek-empty" reason:@"Can not initialize Magtek Device Manager ([DCGUADynamoDeviceManager getAvailableDevices] returns an empty array)." userInfo:nil];
            deviceManager = [[DCGUADynamoDeviceManager alloc] init:[devices objectAtIndex:0]];
            break;
        /* Not implemented for iOS.
        case WL_DEVICE_DC_IDT_UNI_PAY:
            devices = [DCGUnipayDeviceManager getAvailableDevices];
            if(devices==nil||[devices count]==0)
                return nil;
            deviceManager = [[DCGUnipayDeviceManager alloc] init:[devices objectAtIndex:0]];
            break;*/
        /* Not implemented for iOS.
        case WL_DEVICE_DC_MIURA:
            devices = [DCGMiuraDeviceManager getAvailableDevices];
            if(devices==nil)
                @throw [Wl_UserException exceptionWithName:@"miura-null" reason:@"Can not initialize IDTech Miura Device Manager ([DCGMiuraDeviceManager getAvailableDevices] returns null)." userInfo:nil];
            if([devices count]==0)
                @throw [Wl_UserException exceptionWithName:@"miura-empty" reason:@"Can not initialize IDTech Miura Device Manager ([DCGMiuraDeviceManager getAvailableDevices] returns an empty array)." userInfo:nil];
            deviceManager = [[DCGMiuraDeviceManager alloc] init:[devices objectAtIndex:0]];
            break;*/
        /* Not implemented for iOS.
        case WL_DEVICE_DC_PAX:
             devices = [DCGPAXDeviceManager getAvailableDevices];
             if(devices==nil||[devices count]==0)
             return nil;
             deviceManager = [[DCGPAXDeviceManager alloc] init:[devices objectAtIndex:0]];
             break;*/
        case WL_DEVICE_VIRTUAL:
             devices = [DCGVirtualDeviceManager getAvailableDevices];
            if(devices==nil)
                @throw [Wl_UserException exceptionWithName:@"virtual-null" reason:@"Can not initialize Virtual Device Manager ([DCGVirtualDeviceManager getAvailableDevices] returns null)." userInfo:nil];
            if([devices count]==0)
                @throw [Wl_UserException exceptionWithName:@"virtual-empty" reason:@"Can not initialize Virtual Device Manager ([DCGVirtualDeviceManager getAvailableDevices] returns an empty array)." userInfo:nil];
            deviceManager = [[DCGVirtualDeviceManager alloc] init:[devices objectAtIndex:0]];
            break;
        default:
            @throw [Wl_UserException exceptionWithName:@"dc-not-implemented" reason:@"Interface for DirecConnect device is not implemented." userInfo:nil];
    }

    Wl_Pay_Ccr_DirectConnect* o_result = [[Wl_Pay_Ccr_DirectConnect alloc] init];
    o_result->devices = devices;
    o_result->deviceManager = deviceManager;
    o_result->id_device = id_device;
    return o_result;

}

+(NSDictionary*)debugGlobal
{
    NSMutableDictionary* a_debug = [[NSMutableDictionary alloc] init];

    NSArray* devices;

    /*devices = [DCGBTMagDeviceManager getAvailableDevices];
    if(devices==nil)
        [a_debug setObject:@"[null]" forKey:@"[DCGBTMagDeviceManager getAvailableDevices]"];
    else if ([devices count]==0)
        [a_debug setObject:@"[empty array]" forKey:@"[DCGBTMagDeviceManager getAvailableDevices]"];
    else
        [a_debug setValue:[NSNumber numberWithLong:[devices count]] forKey:@"[DCGBTMagDeviceManager getAvailableDevices]"];

    devices = [DCGUniMagDeviceManager getAvailableDevices];
    if(devices==nil)
        [a_debug setObject:@"[null]" forKey:@"[DCGUniMagDeviceManager getAvailableDevices]"];
    else if ([devices count]==0)
        [a_debug setObject:@"[empty array]" forKey:@"[DCGUniMagDeviceManager getAvailableDevices]"];
    else
        [a_debug setValue:[NSNumber numberWithLong:[devices count]] forKey:@"[DCGUniMagDeviceManager getAvailableDevices]"];*/

    devices = [DCGUADynamoDeviceManager getAvailableDevices];
    if(devices==nil)
        [a_debug setObject:@"[null]" forKey:@"[DCGUADynamoDeviceManager getAvailableDevices]"];
    else if ([devices count]==0)
        [a_debug setObject:@"[empty array]" forKey:@"[DCGUADynamoDeviceManager getAvailableDevices]"];
    else
        [a_debug setValue:[NSNumber numberWithLong:[devices count]] forKey:@"[DCGUADynamoDeviceManager getAvailableDevices]"];

    /* Not implemented for iOS.
    devices = [DCGMiuraDeviceManager getAvailableDevices];
    if(devices==nil)
        [a_debug setObject:@"[null]" forKey:@"[DCGMiuraDeviceManager getAvailableDevices]"];
    else if ([devices count]==0)
        [a_debug setObject:@"[empty array]" forKey:@"[DCGMiuraDeviceManager getAvailableDevices]"];
    else
        [a_debug setValue:[NSNumber numberWithLong:[devices count]] forKey:@"[DCGMiuraDeviceManager getAvailableDevices]"];*/

    devices = [DCGVirtualDeviceManager getAvailableDevices];
    if(devices==nil)
        [a_debug setObject:@"[null]" forKey:@"[DCGVirtualDeviceManager getAvailableDevices]"];
    else if ([devices count]==0)
        [a_debug setObject:@"[empty array]" forKey:@"[DCGVirtualDeviceManager getAvailableDevices]"];
    else
        [a_debug setValue:[NSNumber numberWithLong:[devices count]] forKey:@"[DCGVirtualDeviceManager getAvailableDevices]"];

    return a_debug;
}

- (NSDictionary*)debugInfo
{
    NSMutableDictionary* a_debug = [[NSMutableDictionary alloc] init];

    NSMutableArray* a_device = [[NSMutableArray alloc] init];
    for(DCGDevice* device in devices)
    {
        NSMutableDictionary* a_device_item = [[NSMutableDictionary alloc] init];
        [a_device_item setObject:[device Address] forKey:@"address"];
        [a_device_item setObject:[device Name] forKey:@"name"];
        [a_device_item setObject:[device Type] forKey:@"type"];
        [a_device_item setObject:[device description] forKey:@"description"];
        [a_device addObject:a_device_item];
    }

    [a_debug setObject:a_device forKey:@"devices"];
    [a_debug setObject:devices==nil?@"[nil]":[NSNumber numberWithInteger:[devices count]] forKey:@"devices.length"];
    [a_debug setObject:deviceManager==nil?@"[nil]":NSStringFromClass([deviceManager class]) forKey:@"deviceManager.class"];
    [a_debug setObject:[NSNumber numberWithInteger:id_device] forKey:@"id_device"];
    [a_debug setObject:NSStringFromClass([self class]) forKey:@"this.class"];

    if(deviceManager!=nil)
    {
        [a_debug setObject:[NSNumber numberWithBool:[deviceManager isCardInserted]] forKey:@"deviceManager.isCardInserted"];
        [a_debug setObject:[NSNumber numberWithBool:[deviceManager isConnected]] forKey:@"deviceManager.isConnected"];
    }

    if(o_card_last==nil)
    {
        [a_debug setObject:@"[nil]" forKey:@"card"];
    }
    else
    {
        NSMutableDictionary* a_card = [[NSMutableDictionary alloc] init];
        if([o_card_last CardholderName]==nil)
            [a_card setObject:@"[nil]" forKey:@"CardholderName"];
        else
            [a_card setObject:[o_card_last CardholderName] forKey:@"CardholderName"];
        [a_card setObject:[o_card_last DataBlock] forKey:@"DataBlock"];
        [a_card setObject:[NSNumber numberWithInteger:[o_card_last DataType]] forKey:@"DataType"];
        [a_card setObject:[o_card_last ExpDate] forKey:@"ExpDate"];
        [a_card setObject:[o_card_last KSN] forKey:@"KSN"];
        [a_card setObject:[o_card_last PAN] forKey:@"PAN"];
        [a_card setObject:[o_card_last ServiceCode] forKey:@"ServiceCode"];
        [a_card setObject:[o_card_last Track1] forKey:@"Track1"];
        [a_card setObject:[o_card_last Track2] forKey:@"Track2"];
        [a_card setObject:[o_card_last Track3] forKey:@"Track3"];

        [a_debug setObject:a_card forKey:@"card"];
    }

    return a_debug;
}

-(id)init
{
    self = [super init];
    if(self)
    {
        o_card_last = nil;
    }
    return self;
}

- (void)onCardInserted:(DCGCardData *)cardData
{
    [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.onCardInserted]"];
}

- (void)onCardRemoved
{
    [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.onCardRemoved]"];
}

- (void)onCardSwiped:(DCGCardData *)cardData
{
    @try
    {
        o_card_last = cardData;

        [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.cardData]"];

        if(cardData!=nil&&[cardData DataType]!=DCGCardDataType_Nil&&[cardData EncryptionParameters]!=nil)
        {
            NSMutableDictionary* a_card = [[NSMutableDictionary alloc] init];

            NSMutableDictionary* a_encrypt = [[NSMutableDictionary alloc] init];
            [a_encrypt setObject:[cardData DataBlock] forKey:@"DataBlock"];
            [a_encrypt setObject:[[cardData EncryptionParameters] EncryptionType] forKey:@"EncryptionType"];
            [a_encrypt setObject:[[cardData EncryptionParameters] HSMDevice] forKey:@"HSMDevice"];
            [a_encrypt setObject:[cardData KSN] forKey:@"KSN"];
            [a_encrypt setObject:[[cardData EncryptionParameters] TerminalType] forKey:@"TerminalType"];

            [a_card setObject:a_encrypt forKey:@"a_encrypt"];
            [a_card setObject:[cardData PAN] forKey:@"s_number_mask"];
            [a_card setObject:[cardData ExpDate] forKey:@"s_expire"];
            [a_card setObject:[NSString stringWithFormat:@"ios.dc.%@",NSStringFromClass([deviceManager class])] forKey:@"s_device"];
            NSString *s_holder = [cardData CardholderName];
            if(s_holder!=nil)
                [a_card setObject:[s_holder stringByReplacingOccurrencesOfString:@"/" withString:@" "] forKey:@"s_holder"];

            [o_controller fireSwipe:a_card];
        }
        else
        {
            [o_controller fireSwipeError];
        }

        if(id_device!=WL_DEVICE_VIRTUAL)
            [deviceManager acceptCard:@"Swipe Card"];
    }
    @catch(id e)
    {
        [o_controller _exception:e forCommand:nil];
    }
}

- (void)onConnected
{
    [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.onConnected]"];
    if(id_device!=WL_DEVICE_VIRTUAL)
        [deviceManager acceptCard:@"Swipe Card"];
}

- (void)onDisconnected
{
    [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.onDisconnected]"];
}

- (void)onMenuSelected:(int)selection
{
    [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.onMenuSelected]"];
}

- (void)onPINEntered:(DCGPINData *)pinData
{
    [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.onPINEntered]"];
}

- (void)onYNAnswered:(int)response {
    [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.onYNAnswered]"];
}

-(void) startup
{
    [self logInfo:@"[Wl_Pay_Ccr_DirectConnect.startup]"];
    [deviceManager connect:self];
}

-(void)tearDown
{
    if(deviceManager!=nil)
    {
        [deviceManager disconnect];
        deviceManager = nil;
    }
    devices = nil;
}

- (void)testSwipe: (NSDictionary*)a_card
{
    if(deviceManager==nil)
    {
        [self logErrorMessage:@"[Wl_Pay_Ccr_DirectConnect.testSwipe] deviceManager is not initialized."];
        return;
    }

    if(id_device!=WL_DEVICE_VIRTUAL)
    {
        [self logErrorMessage:@"[Wl_Pay_Ccr_DirectConnect.testSwipe] Can not do testSwipe() because current device is not for testing purposes."];
        return;
    }

    [deviceManager acceptCard:@"Enter card data."];
}

@end